File: src/com/aopphp/go/pointcut/OrPointcut.java
package com.aopphp.go.pointcut;

import com.jetbrains.php.lang.psi.elements.PhpClass;
import com.jetbrains.php.lang.psi.elements.PhpClassMember;
import com.jetbrains.php.lang.psi.elements.PhpNamedElement;

import java.util.HashSet;
import java.util.Set;

/**
 * Logic "OR" pointcut implementation
 */
public class OrPointcut extends AndPointcut {


    public OrPointcut(Pointcut first, Pointcut second) {

        this.first = first;
        this.second = second;

        // combine both sets
        Set<KindFilter> combined = new HashSet<KindFilter>(first.getKind());
        combined.addAll(second.getKind());

        this.kind        = combined;
        this.classFilter = new OrPointFilter(first.getClassFilter(), second.getClassFilter());
    }

    @Override
    public boolean matches(PhpNamedElement element) {
        return isMatchesPointcut(element, first) || isMatchesPointcut(element, second);
    }
}

File: src/main/kotlin/com/aopphp/go/pointcut/OrPointcut.kt
package com.aopphp.go.pointcut

import com.jetbrains.php.lang.psi.elements.PhpNamedElement

class OrPointcut(first: Pointcut, second: Pointcut) : AndPointcut(first, second) {
    init {
        _classFilter = OrPointFilter(first.getClassFilter(), second.getClassFilter())
    }

    // Override kind to be the union rather than intersection
    private val _orKind: Set<KindFilter> = first.getKind() + second.getKind()
    override fun getKind() = _orKind

    override fun matches(element: PhpNamedElement) =
        isMatchesPointcut(element, first) || isMatchesPointcut(element, second)
}
